WordPress/Classes/Services/ReaderPostService.m

```diff
@@ -139,23 +137,20 @@ - (void)fetchPost:(NSUInteger)postID forSite:(NSUInteger)siteID isFeed:(BOOL)isF
             return;
         }
 
-        NSManagedObjectContext *context = self.managedObjectContext;
-        ReaderPost *post = [self createOrReplaceFromRemotePost:remotePost forTopic:nil inContext:context];
+        NSManagedObjectID * __block postObjectID = nil;
+        [self.coreDataStack performAndSaveUsingBlock:^(NSManagedObjectContext *context) {
+            ReaderPost *post = [self createOrReplaceFromRemotePost:remotePost forTopic:nil inContext:context];
 
-        NSError *error;
-        BOOL obtainedID = [context obtainPermanentIDsForObjects:@[post] error:&error];
-        if (!obtainedID) {
-            DDLogError(@"Error obtaining a permanent ID for post. %@, %@", post, error);
-        }
-
-        [[ContextManager sharedInstance] saveContext:context];
-        success(post);
-
-    } failure:^(NSError *error) {
-        if (failure) {
-            failure(error);
-        }
-    }];
+            NSError *error;
+            BOOL obtainedID = [context obtainPermanentIDsForObjects:@[post] error:&error];
+            if (!obtainedID) {
+                DDLogError(@"Error obtaining a permanent ID for post. %@, %@", post, error);
+            }
+            postObjectID = post.objectID;
+        } completion:^{
+            success([self.coreDataStack.mainContext existingObjectWithID:postObjectID error:nil]);
+        } onQueue:dispatch_get_main_queue()];
+    } failure:failure];
```

Comment:
They are similar, but this code won't have the same issue of calling both the success and failure blocks. The `failure` block here will be called upon the HTTP API request failure (sent by the "remote" type above). And the `success` block will be called when the HTTP API request succeeds. So, only one of them will be called.

However, the error reported by `obtainPermanentIDsForObjects` was, and still is, ignored here. It would be technically correct to capture the error and call the failure with it. Propogating the error up would only make the code more correct.

But at the same time, I'm not sure what's the possibility of this error and the value of reporting the error upwards when it occurs. And, from my search for this API symbol, I think all call sites ignore this error.

[The API doc](https://developer.apple.com/documentation/coredata/nsmanagedobjectcontext/1506793-obtainpermanentidsforobjects) doesn't say anything about what the error could be. I presume, the corresponding SQLite API reports an error and Core Data thinks it's best to surface it rather than swallow it. And again, this is my assumption, the error might have the same possibility of writing a few bytes into the file system. So, I'm guessing, the chance that this error occurs might be low, and other things may go terrible wrong too (like how the app would handle Core Data error if there is no space available on the device) when this error occurs.

> I checked the docs for [existingObjectWithID:error:](https://developer.apple.com/documentation/coredata/nsmanagedobjectcontext/1506686-existingobjectwithid) but I couldn't find info on what would happen if the NSManagedObjectID input is nil

I believe the function crashes, throwing a "invalid argument" exception or something.

Reply:
Thanks for the details.